File: fraiseql-scala/src/test/scala/com/fraiseql/ExportTypesSpec.scala
package com.fraiseql

import com.fraiseql.schema.Schema
import org.scalatest.BeforeAndAfterEach
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import upickle.default.*
import java.nio.file.Files
import java.nio.file.Paths

/**
 * Tests for minimal types.json export (TOML-based workflow)
 *
 * Validates that Schema.exportTypes() function generates minimal schema
 * with only types (no queries, mutations, observers, security, etc.)
 */
class ExportTypesSpec extends AnyFlatSpec with Matchers with BeforeAndAfterEach {
  override def beforeEach(): Unit = {
    Schema.reset()
  }

  override def afterEach(): Unit = {
    Schema.reset()
  }

  "exportTypes" should "export minimal schema with single type" in {
    // Register a single type
    Schema.registerType("User", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
      "name" -> Map("type" -> "String", "nullable" -> false),
      "email" -> Map("type" -> "String", "nullable" -> false),
    ), Some("User in the system"))

    // Export minimal types
    val json = Schema.exportTypes(pretty = true)
    val parsed = ujson.read(json)

    // Should have types section
    assert(parsed.obj.contains("types"))
    assert(parsed("types").isInstanceOf[ujson.Arr])
    assert(parsed("types").arr.length == 1)

    // Should NOT have queries, mutations, observers
    assert(!parsed.obj.contains("queries"))
    assert(!parsed.obj.contains("mutations"))
    assert(!parsed.obj.contains("observers"))
    assert(!parsed.obj.contains("authz_policies"))

    // Verify User type
    val userDef = parsed("types")(0)
    assert(userDef("name").str == "User")
    assert(userDef("description").str == "User in the system")
  }

  it should "export minimal schema with multiple types" in {
    // Register User type
    Schema.registerType("User", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
      "name" -> Map("type" -> "String", "nullable" -> false),
    ))

    // Register Post type
    Schema.registerType("Post", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
      "title" -> Map("type" -> "String", "nullable" -> false),
      "authorId" -> Map("type" -> "ID", "nullable" -> false),
    ))

    // Export minimal
    val json = Schema.exportTypes(pretty = true)
    val parsed = ujson.read(json)

    // Check types count
    assert(parsed("types").arr.length == 2)

    // Verify both types present
    val typeNames = parsed("types").arr.map(t => t("name").str)
    assert(typeNames.contains("User"))
    assert(typeNames.contains("Post"))
  }

  it should "not include queries in minimal export" in {
    // Register type
    Schema.registerType("User", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
    ))

    // Export minimal
    val json = Schema.exportTypes(pretty = true)
    val parsed = ujson.read(json)

    // Should have types
    assert(parsed.obj.contains("types"))

    // Should NOT have queries
    assert(!parsed.obj.contains("queries"))
    assert(!parsed.obj.contains("mutations"))
  }

  it should "export compact format when pretty is false" in {
    Schema.registerType("User", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
    ))

    // Export compact
    val compact = Schema.exportTypes(false)

    // Should be valid JSON
    val parsed = ujson.read(compact)
    assert(parsed.obj.contains("types"))

    // Compact JSON should be smaller or equal to pretty
    val pretty = Schema.exportTypes(true)
    assert(compact.length <= pretty.length)
  }

  it should "export pretty format when pretty is true" in {
    Schema.registerType("User", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
    ))

    // Export pretty
    val json = Schema.exportTypes(true)

    // Should contain newlines (pretty format)
    assert(json.contains("\n"))

    // Should be valid JSON
    ujson.read(json)
  }

  it should "export types to file" in {
    Schema.registerType("User", Map(
      "id" -> Map("type" -> "ID", "nullable" -> false),
      "name" -> Map("type" -> "String", "nullable" -> false),
    ))

    // Export to temporary file
    val tmpFile = "/tmp/fraiseql_types_test_scala.json"

    // Remove file if exists
    if (Files.exists(Paths.get(tmpFile))) {
      Files.delete(Paths.get(tmpFile))
    }

    // Export to file
    Schema.exportTypesFile(tmpFile)

    // Verify file exists and is valid JSON
    assert(Files.exists(Paths.get(tmpFile)))

    val content = Files.readString(Paths.get(tmpFile))
    val parsed = ujson.read(content)

    assert(parsed.obj.contains("types"))
    assert(parsed("types").arr.length == 1)

    // Cleanup
    Files.delete(Paths.get(tmpFile))
  }

  it should "handle empty schema gracefully" in {
    // Export with no types registered
    val json = Schema.exportTypes(true)
    val parsed = ujson.read(json)

    // Should still have types key (as empty array)
    assert(parsed.obj.contains("types"))
    assert(parsed("types").isInstanceOf[ujson.Arr])
    assert(parsed("types").arr.isEmpty)
  }
}

File: sdks/community/fraiseql-ruby/spec/export_types_spec.rb
require 'spec_helper'
require 'json'

describe 'Export Types - Minimal Schema Export' do
    before(:each) do
      FraiseQL::Schema.reset
    end

    it 'exports minimal schema with single type' do
      # Register a single type
      FraiseQL::Schema.register_type('User', [
        { name: 'id', type: 'ID', nullable: false },
        { name: 'name', type: 'String', nullable: false },
        { name: 'email', type: 'String', nullable: false }
      ], 'User in the system')

      # Export minimal types
      schema_json = FraiseQL::Schema.export_types(true)
      parsed = JSON.parse(schema_json)

      # Should have types section
      expect(parsed).to have_key('types')
      expect(parsed['types']).to be_an(Array)
      expect(parsed['types'].length).to eq(1)

      # Should NOT have queries, mutations, observers, etc.
      expect(parsed).not_to have_key('queries')
      expect(parsed).not_to have_key('mutations')
      expect(parsed).not_to have_key('observers')
      expect(parsed).not_to have_key('authz_policies')

      # Verify User type
      user_type = parsed['types'].first
      expect(user_type['name']).to eq('User')
      expect(user_type['description']).to eq('User in the system')
    end

    it 'exports minimal schema with multiple types' do
      # Register multiple types
      FraiseQL::Schema.register_type('User', [
        { name: 'id', type: 'ID', nullable: false },
        { name: 'name', type: 'String', nullable: false }
      ])

      FraiseQL::Schema.register_type('Post', [
        { name: 'id', type: 'ID', nullable: false },
        { name: 'title', type: 'String', nullable: false },
        { name: 'authorId', type: 'ID', nullable: false }
      ])

      # Export minimal
      schema_json = FraiseQL::Schema.export_types(true)
      parsed = JSON.parse(schema_json)

      # Check types count
      expect(parsed['types'].length).to eq(2)

      # Verify both types present
      type_names = parsed['types'].map { |t| t['name'] }
      expect(type_names).to include('User', 'Post')
    end

    it 'does not include queries in minimal export' do
      # Register type and query
      FraiseQL::Schema.register_type('User', [
        { name: 'id', type: 'ID', nullable: false }
      ])

      # (In a real SDK, queries would be registered here)
      # But minimal export should only have types

      # Export minimal
      schema_json = FraiseQL::Schema.export_types(true)
      parsed = JSON.parse(schema_json)

      # Should have types
      expect(parsed).to have_key('types')

      # Should NOT have queries
      expect(parsed).not_to have_key('queries')
      expect(parsed).not_to have_key('mutations')
    end

    it 'exports compact format when pretty is false' do
      FraiseQL::Schema.register_type('User', [
        { name: 'id', type: 'ID', nullable: false }
      ])

      # Export compact
      schema_json = FraiseQL::Schema.export_types(false)

      # Should be valid JSON
      parsed = JSON.parse(schema_json)
      expect(parsed).to have_key('types')

      # Compact JSON should be smaller than pretty-printed
      compact = FraiseQL::Schema.export_types(false)
      pretty = FraiseQL::Schema.export_types(true)
      expect(compact.length).to be < pretty.length
    end

    it 'exports pretty format when pretty is true' do
      FraiseQL::Schema.register_type('User', [
        { name: 'id', type: 'ID', nullable: false }
      ])

      # Export pretty
      schema_json = FraiseQL::Schema.export_types(true)

      # Should contain newlines (pretty format)
      expect(schema_json).to include("\n")

      # Should be valid JSON
      parsed = JSON.parse(schema_json)
      expect(parsed).to have_key('types')
    end

    it 'exports types to file' do
      FraiseQL::Schema.register_type('User', [
        { name: 'id', type: 'ID', nullable: false },
        { name: 'name', type: 'String', nullable: false }
      ])

      # Export to temporary file
      tmp_file = '/tmp/fraiseql_types_test.json'

      # Remove file if exists
      File.delete(tmp_file) if File.exist?(tmp_file)

      # Export to file
      FraiseQL::Schema.export_types_file(tmp_file)

      # Verify file exists and is valid JSON
      expect(File.exist?(tmp_file)).to be true

      content = File.read(tmp_file)
      parsed = JSON.parse(content)

      expect(parsed).to have_key('types')
      expect(parsed['types'].length).to eq(1)

      # Cleanup
      File.delete(tmp_file)
    end

    it 'handles empty schema gracefully' do
      # Export with no types registered
      schema_json = FraiseQL::Schema.export_types(true)
      parsed = JSON.parse(schema_json)

      # Should still have types key (as empty array)
      expect(parsed).to have_key('types')
      expect(parsed['types']).to be_an(Array)
      expect(parsed['types'].length).to eq(0)
    end
  end
